docs/style-guide.md

```diff
@@ -0,0 +1,439 @@
+
+## Sources
+
+* [Learn Rust](https://doc.rust-lang.org/)
+  * The canonical source for Rust information, but it doesn't offer advice on all aspects of code design.
+* [Introduction - Learning Rust With Entirely Too Many Linked Lists](https://rust-unofficial.github.io/too-many-lists/)
+  * This is brilliantly written and very educational about how you get into a Rust mindset.
+* [Elegant Library APIs in Rust](https://deterministic.space/elegant-apis-in-rust.html)
+  * This has a lot of good points and is well worth a read, but be warned that some of the details about implementation are somehwat out of date (2017). It has a video too.
+* [Good Practices for Writing Rust Libraries](https://pascalhertleif.de/artikel/good-practices-for-writing-rust-libraries/)
+  * Shorter and more focussed on the act of coding the libraries, and less about API design. Also potentially out of date in places.
+* [Strategies for Returning References in Rust](http://bryce.fisher-fleig.org/blog/strategies-for-returning-references-in-rust/index.html)
+  * Though I don't believe we should be doing this, it's still an interesting read.
+
+## Other Useful Links
+
+* Write and run Rust snippets: https://play.rust-lang.org
+  * You can save snippets in permantent links and incluce them as working examples in docs.
+* Write code and see what it compiles to: https://rust.godbolt.org
+  * Note that you need to enable compiler optimizations via `-C opt-level=3` if you are looking to meaningfully compare two code snippets.
+
+# Naming Conventions
+
+### No Special Naming :: required
+
+Follow the naming advice in [Naming - Rust API Guidelines](https://rust-lang.github.io/api-guidelines/naming.html) at all times, and use [Rust RFC 0430](https://github.com/rust-lang/rfcs/blob/master/text/0430-finalizing-naming-conventions.md) for any ambiguous issues (e.g. Crate naming).
+
+Based on [this comment](https://github.com/rust-lang/api-guidelines/issues/29#issuecomment-342745898) I think the advice for Crate names should be:
+
+* Infer Crate names from package names where feasible:
+  * A package is the directory in which the Cargo.toml exists.
+* Prefer single (lowercase) words for package names where feasible
+* Use __lowercase with hyphen__ to split words in package names if needed
+
+This means that while Crate names are not trivially 
+I believe this would result in Crate names (which must be Rust identifiers) being identical to the package name at all times.
+
+## Module Layout
+
+Modules form the basic units of imported code in Rust, so identifying where to split functionality and what to call modules is essential. I suspect this is the right level to divide things like collation vs segmentation etc. but it's also far too early to make hard recommendations about this, so I proposed to leave this issue open for the time being (moving code between modules later should be relatively easy).
```

Comment:
Another "I" statement.
I'd remove that paragraph as its inconclusive and at the moment doesn't help a newcomer work with the project. We'll add it once we establish the norms